### PR TITLE
Bump to latest libkv + dependencies

### PR DESCRIPTION
**Pull request by dhiltgen**
The latest libkv uses a different etcd library.  Unfortunately
that library uses some funky import paths, so I've added a new cleanup
routine for our vendor scripts to be able to normalize the imports
to be consistent with how imports work in this tree.

Signed-off-by: Daniel Hiltgen daniel.hiltgen@docker.com

### CODE REVIEW

**Comment by icecrime** [comment]
To summarize the discussions here:
- The `etcd` client library we currently use is deprecated and has known issues (I'm told potential crashes due to lack of thread-safety).
- This PR proposes to upgrade to the newer, supported, `etcd` client library, which only drawback is to pull a 40k+ LoC dependency with it ([`https://github.com/ugorji/go`](https://github.com/ugorji/go)).
- Most likely, the future `etcd` API v3 will be protobuf based (a dependency we already have), won't require this anymore, _and_ should be available in time for Docker 1.10.

That is to say that I feel the current choice is:
1. Merge this PR in order to use the more robust client package, and close our eyes on the amount of code it requires. Most likely, this will all be cleaned up for 1.10 with the newer package.
2. Don't merge this PR, knowing that the daemon might crash when `--cluster-store=etcd://` is used, and wait for `etcd` API v3 to upgrade.

I vote option 1.

**Comment by jfrazelle** [comment]
:see_no_evil: 

no. 1

**Comment by icecrime** [comment]
Ping @abronan @mavenugo.

**Comment by abronan** [comment]
Thanks for summing up the situation @icecrime 

:see_no_evil: option 1

**Comment by jfrazelle** [comment]
are we all closing out eye hahahaha lol

**Comment by nathanleclaire** [comment]
:see_no_evil:

**Comment by icecrime** [comment]
Discussed with @mavenugo and team: they will be running stress tests tomorrow on Docker daemon with multihost networking backed by etcd using the existing driver. If we do experience the kind of crashes that we worry might happen, we will consider taking this PR for 1.9. If we don't, it might not be worth the effort.

**Comment by vdemeester** [comment]
:see_no_evil: :hear_no_evil:

File: hack/.vendor-helpers.sh [hunk 1]
```diff
@@ -128,3 +128,12 @@ clean() {
 
 	echo done
 }
+
+fix_rewritten_imports () {
```
Comment (tiborvass) [inline hack/.vendor-helpers.sh]: @dhiltgen a comment explaining what this function is doing would be appreciated :)

Comment (dhiltgen) [inline hack/.vendor-helpers.sh]: will do

File: hack/.vendor-helpers.sh [hunk 2]
```diff
@@ -128,3 +128,13 @@ clean() {
...
 }
+
+# Fix up hard-coded imports that refer to Godeps paths so they'll work with our vendoring
+fix_rewritten_imports () {
+	local pkg="$1"
+	local remove="${pkg}/Godeps/_workspace/src/"
+	local target="vendor/src/$pkg"
+
+	echo "$pkg: fixing rewritten imports"
+	find "$target" -name \*.go -exec sed -i -e "s|\"${remove}|\"|g" {} \;
```
Comment (tiborvass) [inline hack/.vendor-helpers.sh]: should be `sed -i''` to make the script work both on osx and linux.

Comment (tianon) [inline hack/.vendor-helpers.sh]: Windows will balk on the "find" :wink:

**Comment by mavenugo** [comment]
@abronan @dhiltgen I tried to integrate this PR on top of #16910 to make sure things work as expected. but I see the dreaded error - 
`Error response from daemon: failed to update store for object type *libnetwork.network: 104: Not a directory (/docker/network/v1.0/network) [23]`

I know @abronan tried to solve it prior to the watch related changes and eventually was closed (https://github.com/docker/libnetwork/pull/582). Do we need that patch in libnetwork in order to make progress ?
I will continue to dig into this. But we need to make sure this works so that we get to test it e2e.

**Comment by mrjana** [comment]
@mavenugo I might know what the problem is. I'll add an e2e integration test in libnetwork to verify etcd to fix this once for all. There is a small change needed in libnetwork to make this work. With that change you might be able to verify the new etcd client in docker.

**Comment by abronan** [comment]
Issue described above fixed in libnetwork with @mavenugo. So we good.

**Comment by mavenugo** [comment]
@mrjana @abronan thanks for the help. Yes, we have identified a fix. there is an sequencing issue with the PRs though. Lets vendor in libnetwork via #16910 and rebase this PR on top of it.
Sounds good ?

**Comment by dhiltgen** [comment]
Just bumped to the latest libkv at @mavenugo's request

**Comment by dhiltgen** [comment]
I have a version of this change rebased on top of 16910 which I'll push up as soon as 16910 is merged.

**Comment by tiborvass** [comment]
@dhiltgen is `7078e7d` the one you pushed after 16910 was merged?

**Comment by dhiltgen** [comment]
correct - it's rebased on top of madhu's change, and includes the latest libnetwork

File: hack/vendor.sh [hunk 3]
```diff
@@ -20,18 +20,20 @@ clone git github.com/tchap/go-patricia v2.1.0
 clone git golang.org/x/net 3cffabab72adf04f8e3b01c5baf775361837b5fe https://github.com/golang/net.git
 
 #get libnetwork packages
-clone git github.com/docker/libnetwork 0521fe53fc3e7d4e7c2e463800f36662c9169f20
+clone git github.com/docker/libnetwork dd1c5f0ffe7697f75a82bd8e4bbd6f225ec5e383
```
Comment (tiborvass) [inline hack/vendor.sh]: @dhiltgen why do we need to bump libnetwork again?

Comment (abronan) [inline hack/vendor.sh]: @tiborvass This includes a fix for `etcd` (see [this comment](https://github.com/docker/docker/pull/16863#issuecomment-147682818))

**Comment by abronan** [comment]
Still LGTM

**Comment by tiborvass** [comment]
LGTM for `hack/`

**Comment by abronan** [comment]
ping @madhu @icecrime

**Comment by mavenugo** [comment]
tested minimal etcd integration with libnetwork with the patch here. LGTM.
once we have proper etcd integration tests, we will cover many cases.

this PR is good to go.

**Pull request merged**